.upstream-tests/test/cuda/proclaim_return_type.pass.cpp

```diff
@@ -0,0 +1,65 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++98, c++03
+
+#include <cuda/functional>
+
+#include <assert.h>
+
+template <class T, class Fn, class... As>
+__host__ __device__
+void test_proclaim_return_type(Fn&& fn, T expected, As... as)
+{
+  {
+    auto f = cuda::proclaim_return_type<T>(cuda::std::forward<Fn>(fn));
+
+    assert(f(as...) == expected);
+    assert(cuda::std::move(f)(as...) == expected);
+  }
+  {
+    const auto f = cuda::proclaim_return_type<T>(fn);
+
+    assert(f(as...) == expected);
+    assert(cuda::std::move(f)(as...) == expected);
+  }
+}
+
+struct callable
+{
+  __host__ __device__ int operator()() const& { return 42; }
+  __host__ __device__ int operator()() const&& { return 42; }
+};
```

Comment:
Just to be pedantic, we should add a `__device__` only one and a `__host__` only one as well.